src/main/rx/rx.c

```diff
@@ -865,9 +865,19 @@ bool isRssiConfigured(void)
 
 bool rxGetFrameDelta(timeDelta_t *deltaUs)
 {
-    if (rxRuntimeState.rcFrameDeltaFn) {
-        *deltaUs = rxRuntimeState.rcFrameDeltaFn();
-        return true;
+    static timeUs_t previousFrameTimeUs = 0;
+    bool result = false;
+
+    *deltaUs = 0;
+    if (rxRuntimeState.rcFrameTimeUsFn) {
+        const timeUs_t frameTimeUs = rxRuntimeState.rcFrameTimeUsFn();
+        if (frameTimeUs) {
+            if (previousFrameTimeUs) {
+                *deltaUs = cmpTimeUs(frameTimeUs, previousFrameTimeUs);
+                result = true;
```

Comment:
This fallback still does not make any sense to me. If we want to report a debug stream that is made up from data with different sources / meanings, can we at least set the MSB of the stream according to the source, so that users get any use of the data?

Reply:
As I mentioned previously, the purpose of the frame rate delta is not for debugging - it's used to drive the RC smoothing auto-cutoffs and also now used in interpolated feedforward processing (and originally for RC interpolation smoothing). Having it available in a debug mode is just a secondary use originally to validate the auto-training behavior of RC Smoothing. It's just happened to turn out to be more useful than originally planned as we've discovered various unexpected behaviors by different receiver protocols. Also as mentioned, the goal at this point is to preserve the existing, tested, and working behavior. This behavior is to fallback to the "forced" update if no RX frame is received. This behavior was inadvertently broken by the first iteration of implementing protocol-level frame rate measurement. I don't disagree that there may be opportunity to so something "better" (like disconnecting the "forced 30ms max update" from the RX processing), but that's a bigger project that should stand on its own as separate work. So again, this PR is just to fix a bug introduced in the earlier improvements on edge cases where the receiver hasn't yet started, or has stopped sending frames.

Reply:
> As I mentioned previously, the purpose of the frame rate delta is not for debugging - it's used to drive the RC smoothing auto-cutoffs and also now used in interpolated feedforward processing (and originally for RC interpolation smoothing). 

Ok, can we then just stop filling these arbitrarily made up values into `debug` variables? Logging something that we know to be phony and useless seems to be a recipe for bug reports by users trying to actually use them for debugging.

I am happy to add code to log actual RX statistics when I get around to it.

Reply:
The logging is under `RC_SMOOTHING_RATE` and it is exactly what is needed for debugging the auto-training of the **RC smoothing** cutoffs. This is what it was designed for and it works perfectly. It also just happens to be useful for debugging various receiver protocols and has started to be used for that (enabled finding multiple bugs). So it's not "phony and useless" - it does everything we need it to do. There has never been a bug report from a user related to this debugging giving false or misleading information.

For example by using this debug value I see that there's a previously unknown problem with SPI receivers in that they're reporting frame rates jumping all over the place and this is affecting the RC smoothing code. This is unrelated to the EXTI unification (but the problem was there as well).

If you want to enable some other receiver protocol debugging I'm fine with that. But there's no use for a value telling when the last frame was received and having that grow indefinitely if the frames stop. After 0.2 seconds we're going to enter stage 1 failsafe anyway.

Reply:
> The logging is under RC_SMOOTHING_RATE and it is exactly what is needed for debugging the auto-training of the RC smoothing cutoffs. This is what it was designed for and it works perfectly. It also just happens to be useful for debugging various receiver protocols and has started to be used for that (enabled finding multiple bugs). So it's not "phony and useless" - it does everything we need it to do. There has never been a bug report from a user related to this debugging giving false or misleading information.

I think that just the same as what we have said in earlier cases where made up 'debug' values were proposed for the sole purpose of allowing some kind of tuning, the same should apply here: If we need amalgamated values for this then this should be solved on the client side, and not in the firmware. Logging proper raw data that the user can make sense of is the right way to go here as well as in other places.
If you are worried about the lack of bug reports to keep track of this I am happy to open one for this purpose.

> For example by using this debug value I see that there's a previously unknown problem with SPI receivers in that they're reporting frame rates jumping all over the place and this is affecting the RC smoothing code. 

Is there any reason that you would not have been able to gain the same insights if the debug data was reported properly, in separate channels for input based and for task based intervals?

> But there's no use for a value telling when the last frame was received and having that grow indefinitely if the frames stop. After 0.2 seconds we're going to enter stage 1 failsafe anyway.

As I stated before, there is just as little value in reporting task based timings in the case that we are in failsafe, since the RC data will be ignored in this case.
However, I think there is value in being able to track data input intervals in the span between 30 ms when the RC task is auto-triggered and 200ms (when we assume that we've lost control of the craft). Force overriding the actual data input interval in this span makes it impossible to debug any dropouts of these durations.